src/request_handler.rs

```diff
@@ -166,27 +176,108 @@ pub(crate) mod tests {
       std::fs::remove_dir(www).unwrap();
-      reqwest::get(format!("http://localhost:{}", port))
-        .await
-        .unwrap();
+      reqwest::get(url).await.unwrap();
     });
     assert_contains(&stderr, "cannot access `www`");
     assert_contains(&stderr, "Caused by:");
   }
+
+  async fn get_html(url: &Url) -> Html {
+    Html::parse_document(
+      &reqwest::get(url.clone())
+        .await
+        .unwrap()
+        .text()
+        .await
+        .unwrap(),
+    )
+  }
+
+  fn css_select<'a>(html: &'a Html, selector: &'a str) -> Vec<ElementRef<'a>> {
+    let selector = Selector::parse(selector).unwrap();
+    html.select(&selector).collect::<Vec<_>>()
+  }
+
+  #[test]
+  fn listing_contains_file() {
+    test(|url, dir| async move {
+      std::fs::write(dir.join("www").join("some-test-file.txt"), "").unwrap();
+      let haystack = get_html(&url).await.root_element().html();
+      let needle = "some-test-file.txt";
+      assert_contains(&haystack, needle);
+    });
+  }
+
+  #[test]
+  fn listing_contains_multiple_files() {
+    test(|url, dir| async move {
+      let www = dir.join("www");
+      std::fs::write(www.join("a.txt"), "").unwrap();
+      std::fs::write(www.join("b.txt"), "").unwrap();
+      let haystack = get_html(&url).await.root_element().html();
+      assert_contains(&haystack, "a.txt");
+      assert_contains(&haystack, "b.txt");
+    });
+  }
+
+  #[test]
+  fn listed_files_are_download_links() {
+    test(|url, dir| async move {
+      std::fs::write(dir.join("www").join("some-test-file.txt"), "").unwrap();
+      let html = get_html(&url).await;
+      guard_unwrap!(let &[a] = css_select(&html, "a").as_slice());
+      assert_eq!(a.inner_html(), "some-test-file.txt");
+      assert_eq!(dbg!(a.value().attr("download")), Some(""));
+    });
+  }
+
+  #[test]
+  fn listed_files_can_be_downloaded() {
+    test(|url, dir| async move {
+      std::fs::write(dir.join("www").join("some-test-file.txt"), "contents").unwrap();
+      let html = get_html(&url).await;
+      guard_unwrap!(let &[a] = css_select(&html, "a").as_slice());
+      let file_url = a.value().attr("href").unwrap();
+      let file_url = dbg!(url.join(file_url).unwrap());
+      let file_contents = reqwest::get(file_url).await.unwrap().text().await.unwrap();
+      assert_eq!(file_contents, "contents");
+    });
+  }
+
+  #[test]
+  fn disallow_access_outside_of_www() {}
+
+  #[test]
+  fn disallow_parent_path_component() {
+    // todo: This is tricky because `Url::from_str` will remove `..` path
+    // components, so we can't use `reqwest`, or anything that uses `Url`
+    // to exercise a get of paths containing `..`
+  }
+
+  #[test]
+  fn disallow_empty_path_component() {
+    test(|url, _dir| async move {
+      assert_eq!(
+        reqwest::get(format!("{}//bar.txt", url))
+          .await
+          .unwrap()
+          .status(),
+        // TODO: This should return `400`
+        StatusCode::INTERNAL_SERVER_ERROR
```

Comment:
We should switch to an Error enum with a `status` method that returns the HTTP status code to return.